packages/template-retail-react-app/app/hooks/use-current-basket.js

```diff
@@ -4,52 +4,40 @@
  */
-import {useCustomerBaskets, useCustomerId, useProducts} from 'commerce-sdk-react-preview'
+import {useCustomerBaskets, useShopperBasketsMutation} from 'commerce-sdk-react-preview'
+import {useCurrentCustomer} from './use-current-customer'
 
+const onClient = typeof window !== 'undefined'
 /**
  * This hook combine some commerce-react-sdk hooks to provide more derived data for Retail App baskets
  * @param id - basket id to get the current used basket among baskets returned, use first basket in the array if not defined
  * @param shouldFetchProductDetail - boolean to indicate if the baskets should fetch product details based on basket items
  */
-export const useCurrentBasket = ({id = '', shouldFetchProductDetail = false} = {}) => {
-    const customerId = useCustomerId() || ''
-    const {
-        data: basketsData,
-        isLoading: isBasketsLoading,
-        error: basketsError
-    } = useCustomerBaskets({parameters: {customerId}})
-    // if id is not defined, by default use the first basket in the list
-    const basket =
-        basketsData?.baskets?.find((basket) => basket.basketId === id) || basketsData?.baskets?.[0]
-    const productIds = basket?.productItems?.map(({productId}) => productId).join(',') ?? ''
-    const {
-        data: products,
-        isLoading: isProductsLoading,
-        error: productsError
-    } = useProducts(
-        {
-            parameters: {
-                ids: productIds,
-                allImages: true
-            }
-        },
+export const useCurrentBasket = ({id = ''} = {}) => {
+    const createBasket = useShopperBasketsMutation('createBasket')
+    const {data: customer} = useCurrentCustomer()
+    const {customerId} = customer
+    const {data: basketsData, ...restOfQuery} = useCustomerBaskets(
+        {parameters: {customerId}},
         {
-            enabled: shouldFetchProductDetail && !!productIds,
-            select: (result) => {
-                // Convert array into key/value object with key is the product id
-                return result?.data?.reduce((result, item) => {
-                    const key = item.id
-                    result[key] = item
-                    return result
-                }, {})
+            enabled: !!customerId && onClient,
+            onSuccess: (data) => {
+                if (!data.total) {
+                    console.log('creating basket')
+                    createBasket.mutate({
+                        body: {}
+                    })
+                }
             }
         }
     )
 
+    // if id is not defined, by default use the first basket in the list
+    const basket =
+        basketsData?.baskets?.find((basket) => basket.basketId === id) || basketsData?.baskets?.[0]
+
     return {
-        error: basketsError || productsError,
-        isLoading: isBasketsLoading || isProductsLoading,
-        products,
+        ...restOfQuery,
```

Comment:
_(GitHub doesn't allow me to comment on unchanged lines 😢  this comment is for the entire return value)_

The way this object is constructed will mix data and query state/methods. If I console log the return value of this hook, it looks like:

<img width="658" alt="Screenshot 2023-03-09 at 12 20 34 PM" src="https://user-images.githubusercontent.com/10948652/224146312-9bd56d33-c181-42e5-a317-309e9f0808c0.png">

I feel like we should not mix basket data with query states.

Reply:
My suggestion:

```js
    const value = {
        ...query,
        data: basket
    }
```

Now, next question is, what about the derived data? 🤔  Some bad ideas:

1. blend it into `data`, this is not ideal because developers does not expect the data type to be different from the API spec
2. remove derived data until we have a better pattern for it, why? because

> prefer duplication over the wrong abstraction

Reply:
At the end of the day, what we want is for developers to use `useCurrentBasket` easily. We recommend react-query and we should stick to its API to make retail-react-app easy to learn for developers.

Reply:
> remove derived data until we have a better pattern for it, why? because
If we do this, there is no point of having this hook. We want to pick a basket so we do not want to keep running 
`basketsData?.baskets?.find((basket) => basket.basketId === id) || basketsData?.baskets?.[0]` all the time, and other derived data

If we were to do this, then there is no point for creating wishlist hook either, we can keep repeating these logic everywhere